### PR TITLE
feat(file): add a visual dropzone

### PR DESCRIPTION
**Pull request by Kiarokh**
fix: https://github.com/Lundalogik/crm-feature/issues/1800

## Review:
- [x] Commits are [atomic](https://seesparkbox.com/foundry/atomic_commits_with_git)
- [x] Commits have the correct *type* for the changes made
- [x] Commits with *breaking changes* are marked as such

### Browsers tested:
(Check any that applies, it's ok to leave boxes unchecked if testing something didn't seem relevant.)

Windows:
- [ ] Chrome
- [ ] Edge
- [ ] Firefox

Linux:
- [ ] Chrome
- [ ] Firefox

macOS:
- [x] Chrome
- [ ] Firefox
- [ ] Safari

Mobile:
- [x] Chrome on Android
- [x] iOS

### CODE REVIEW

File: src/components/file/examples/file-no-dropzone.tsx
Comment (Befkadu1) [inline src/components/file/examples/file-no-dropzone.tsx:11]: It is good to mention that `readonly` disables adding and removing files.

**Review by Befkadu1 — CHANGES_REQUESTED** [review]

**Comment by Befkadu1** [comment]
@adrianschmidt can you look this PR a bit if you have time especially the styling part?

File: src/translations/sv.ts
Comment (adrianschmidt) [reply src/translations/sv.ts:7]: Good catch!

File: src/components/file/examples/file-no-dropzone.tsx
Comment (adrianschmidt) [reply src/components/file/examples/file-no-dropzone.tsx:11]: Hmm… I'm not sure, but I think I disagree. For one thing, I think it ought to be fairly self-evident that files can't be added and removed when the component is `readonly`. But, even if that _does_ need explaining, I think it should be explained in the docs for `readonly`, or in an example specifically for `readonly`. I don't think that it should be explained here, in the example for `dropzone`.

File: src/components/file/file.tsx
Comment (adrianschmidt) [inline src/components/file/file.tsx:104]: Do we really need this option? Can't we at least wait to introduce it until someone asks for it?

File: src/translations/sv.ts
Comment (Kiarokh) [reply src/translations/sv.ts]: indeed. Just made a fixup

File: src/components/chip-set/chip-set.tsx
Comment (Kiarokh) [reply src/components/chip-set/chip-set.tsx:236]: > I think it is possible to remove the constructor() by changing the binded functions to arrow function as well but it is not related to this PR.

That's right. We can perhaps do it later in another PR ;)

Comment (Kiarokh) [reply src/components/chip-set/chip-set.tsx:236]: > What if we refactor it like this, then we can remove componentWillLoad function.

I don't know. Do you want me to change it this way? Or should we also improve this in the PR that we change the bound functions to arrow functions?

File: src/components/file/file.tsx
Comment (Kiarokh) [reply src/components/file/file.tsx:104]: I think they already did when I demoed it. Example was when the component is used on a phone, they cannot drag and drop. So if someone wants to optionally hide this (like when they have a solution made for phones, or when god-knows-why) then they can do so programatically.

File: src/components/file/examples/file-no-dropzone.tsx
Comment (Kiarokh) [reply src/components/file/examples/file-no-dropzone.tsx:11]: Me neither. I think the concept comes with `readonly`. So I wouldn't explain it.

File: src/components/chip-set/chip-set.tsx
Comment (Befkadu1) [reply src/components/chip-set/chip-set.tsx:236]: Yes we can do that in this PR.

File: src/components/file/file.tsx
Comment (Kiarokh) [reply src/components/file/file.tsx:104]: it'll also prevent complains from people who don't like the change :P

**Review by Befkadu1 — DISMISSED** [review]
It is fine for me. @adrianschmidt we will merge it if you are ok with the PR

File: src/components/file/file.tsx
Comment (adrianschmidt) [reply src/components/file/file.tsx:104]: Ah… ok… hmm… 🤔

I rather think it would be better to build in feature-detection for drag-and-drop and not show this if drag-and-drop isn't supported. But that's more complexity that I also think we shouldn't do unless there's explicit demand for it.

I still suggest that we remove this feature until it's explicitly asked for. It's easy to add later, but harder to remove if we realise there's a better way to do it.

Second opinion @jgroth?

Comment (adrianschmidt) [reply src/components/file/file.tsx:104]: Just "requesting changes" until we've decided this one (and I've looked over the css like Bef asked me to).

**Review by adrianschmidt — CHANGES_REQUESTED** [review]

File: src/components/file/examples/file-accepted-types.tsx
Comment (adrianschmidt) [inline src/components/file/examples/file-accepted-types.tsx:4]: We're not limiting the file extension, we're limiting the file type. It seems like at least macOS is deriving the file type from the file extension, so the effect on macOS is that the file extension is limited, but I'm not sure that's true for all systems.

File: src/components/file/file.tsx
Comment (jgroth) [reply src/components/file/file.tsx:104]: Yeah I think I agree, it should be quite easy to add this later if it is needed :slightly_smiling_face: I general we should try and be a bit restrictive of what we add to the public interface since it always easier to add stuff than change or remove them

File: src/components/chip-set/partial-styles/_file-picker.scss
Comment (adrianschmidt) [inline src/components/chip-set/partial-styles/_file-picker.scss:26]: We might need to limit the max-width of the label too:

image

File: src/components/chip-set/partial-styles/_file-picker.scss
Comment (adrianschmidt) [inline src/components/chip-set/partial-styles/_file-picker.scss:20]: Nice 😁

File: src/components/input-field/partial-styles/_readonly.scss
Comment (adrianschmidt) [inline src/components/input-field/partial-styles/_readonly.scss:34]: > prevent leading icons becoming transparent in `readonly`

Hmm… this seems like something that ought to be fixed in chip-set rather than here? 🤔

File: src/components/chip-set/chip-set.tsx
Comment (adrianschmidt) [inline src/components/chip-set/chip-set.tsx:701]: This change is very definitely not a `feat`. It's a `refactor`. There's no reason we should trigger the release of a new minor version because of this change, and the change should not be documented as a "new feature" in the changelog.

File: src/components/input-field/partial-styles/_readonly.scss
Comment (Kiarokh) [reply src/components/input-field/partial-styles/_readonly.scss:34]: I think it's fine here. Because we don't want the leading or trailing icons to become transparent and look disabled in `readonly` mode. It's also true in input field.

File: src/components/chip-set/chip-set.tsx
Comment (Kiarokh) [reply src/components/chip-set/chip-set.tsx:701]: That's right. This was a 🧠💨

File: src/components/file/examples/file-accepted-types.tsx
Comment (Kiarokh) [reply src/components/file/examples/file-accepted-types.tsx:4]: OK I'll revert back

File: src/components/file/file.tsx
Comment (Kiarokh) [reply src/components/file/file.tsx:104]: OK. I'll remove it

File: src/components/input-field/partial-styles/_readonly.scss
Comment (adrianschmidt) [reply src/components/input-field/partial-styles/_readonly.scss:34]: Yeah, but if we fix it here, it doesn't apply to chip-set in readonly? If we fix it in chip-set, it should automatically apply in all components using chip-set.

File: src/components/chip-set/partial-styles/_file-picker.scss
Comment (Kiarokh) [reply src/components/chip-set/partial-styles/_file-picker.scss]: this commit f7ba8c7c207815b4b3843e5fc96ffb37b8ce9b61 (fix(chip-set): make long labels truncate correctly when there is leading icon) was needed to make this work. The fixup after takes care of label width

**Comment by Kiarokh** [comment]
@adrianschmidt I'm done with your requested changes

File: src/components/file/file.scss
Comment (adrianschmidt) [inline src/components/file/file.scss:4]: This doesn't seem to be used anywhere?

Comment (Kiarokh) [reply src/components/file/file.scss:4]: Oh good catch. It's just trash, left from my first version of the dropzone, when dropzone was visually separated from the input field.

Comment (Kiarokh) [reply src/components/file/file.scss:4]: i'll fixup

File: src/components/file/file.tsx
Comment (adrianschmidt) [inline src/components/file/file.tsx:181]: I added some interactivity to the styles 🙂

https://user-images.githubusercontent.com/154725/126768512-9a88a5d2-71c3-4693-8625-36a1b74337ae.mov

Comment (adrianschmidt) [reply src/components/file/file.tsx:181]: There may be more cool things we can do with changing the pointer and stuff, but I don't have time before my vacation 😅

**Review by adrianschmidt — APPROVED** [review]
👏

File: src/components/file/file.tsx
Comment (Kiarokh) [reply src/components/file/file.tsx:181]: Lovely, so you basically wanna close this issue: https://github.com/Lundalogik/crm-feature/issues/2249 😄 in the same time

Comment (Kiarokh) [reply src/components/file/file.tsx:181]: Wh have [an issue](https://github.com/Lundalogik/crm-feature/issues/2249) for this. It can wait for perfection after the vacation ;)

Comment (adrianschmidt) [reply src/components/file/file.tsx:181]: Ah, yeah, I guess it would be really nice to have _all_ drop zones light up a bit when the user drags something anywhere on the page, and then still have each individual drop zone become even more highlighted when the user drags over it.

**Pull request merged**
